common/src/state/data_transfer.rs

```diff
@@ -0,0 +1,258 @@
+        TransferState {
+            inner: Arc::new(Mutex::new(TransferStates::default())),
+        }
+    }
+
+    pub async fn matches(&self, state: TransferStates) -> bool {
+        *self.inner.lock().await == state
+    }
+
+    pub async fn update(&self, cancel: bool) {
+        let mut v = self.inner.lock().await;
+        *v = v.swap(cancel);
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TransferProgress {
+    Starting,
+    Progress(u8),
+    Finishing,
+    Paused(u8),
+    Cancelling,
+    Error,
+}
+
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub enum TransferStates {
+    #[default]
+    Normal,
+    Cancel,
+    Pause,
+}
+
+impl TransferStates {
+    pub fn swap(&self, cancel: bool) -> TransferStates {
+        match self {
+            TransferStates::Normal => {
+                if cancel {
+                    TransferStates::Cancel
+                } else {
+                    TransferStates::Pause
+                }
+            }
+            TransferStates::Cancel => TransferStates::Cancel,
+            TransferStates::Pause => {
+                if cancel {
+                    TransferStates::Cancel
+                } else {
+                    TransferStates::Normal
+                }
+            }
+        }
+    }
+}
+
+unsafe impl Send for TransferStates {}
+
+#[derive(Debug, Clone)]
+pub enum TrackerType {
+    FileUpload,
+    FileDownload,
+}
+
+#[derive(Debug, Clone)]
+pub struct FileProgress {
+    // Use an uuid for duplicate file names
+    pub id: Uuid,
+    pub file: String,
+    pub progress: TransferProgress,
+    pub description: Option<String>,
+    // Flag used to pause or cancel this transfer
+    pub state: TransferState,
+}
+
+impl PartialEq for FileProgress {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+            && self.file == other.file
+            && self.progress == other.progress
+            && self.description == other.description
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct TransferTracker {
+    pub file_progress_upload: Vec<FileProgress>,
+    pub file_progress_download: Vec<FileProgress>,
+}
+
+impl TransferTracker {
+    pub fn start_file_transfer(
+        &mut self,
+        id: Uuid,
+        file: String,
+        state: TransferState,
+        tracker: TrackerType,
+    ) {
+        match tracker {
+            TrackerType::FileUpload => self.file_progress_upload.push(FileProgress {
+                id,
+                file,
+                progress: TransferProgress::Starting,
+                description: None,
+                state,
+            }),
+            TrackerType::FileDownload => self.file_progress_download.push(FileProgress {
+                id,
+                file,
+                progress: TransferProgress::Starting,
+                description: None,
+                state,
+            }),
+        }
+    }
+
+    pub fn update_file_upload(
+        &mut self,
+        file_id: &Uuid,
```

Comment:
```suggestion
        file_id: Uuid,
```

Reply:
isnt just using refs better if it can be? reduces clones.

Reply:
It is better to borrow when possible, but since `Uuid` derives `Copy`, it is inexpensive to do